### PR TITLE
AWS Cloud9 IDE Environment Setup Instructions

### PR DESCRIPTION
**Pull request by cindyledev**
## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [ ] **New Feature**: Change which adds functionality
- [x] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

I thought this would be nice to have for Windows users who are interested in developing Telescope using a cloud development environment.

## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [x] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

File: env.example [hunk 1]
```diff
@@ -15,6 +18,8 @@ PORT=3000
 #   API_URL=https://dev.telescope.cdot.systems
 #   (our production server)
 #   API_URL=https://telescope.cdot.systems
+#   (AWS Cloud9 development)
+#   API_URL=${MY_PUBLIC_IP}:3000
```
Comment (cindyledev) [inline env.example:22]: I'm having a problem with storing the public IP address as a variable in .env. 

If I hardcode the IP address like
`API_URL=http://35.174.16.133:3000`
the frontend will pick up the feeds from the backend with no problems but I don't want to hardcode it

The public IP address of the EC2 instance will change when it shuts down so I want it to have its own env variable.  Anyone know how I can do this?

File: docs/aws-cloud9.md [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+## Now to get started with development...
+
+Run `npm install` to install all dependencies
+
+You may need up to four terminals, remember to run them in the `telescope directory`
+
+1. `redis-server` for Redis
+2. `sudo /etc/init.d/elasticsearch start` for Elasticsearch
+3. `cd src/frontend`
```
Comment (cindyledev) [inline docs/aws-cloud9.md:156]: I forgot `PORT=3000 npm start` here

**Review by humphd — CHANGES_REQUESTED** [review]
This is really cool!

One thing I think it needs is more help giving some sense of cost.  You have some good warnings, but can we do better?  What would it cost to run this for a day, a week, a month, a term?  Could you use the credits from https://education.github.com/pack to cover it?  How to you ensure that it's 100% "off" and not costing you anything when you're done.

I fear sending students toward AWS without really understanding the cost.  But if they do, this will be very cool.

**Comment by humphd** [comment]
Seneca is now an AWS Educate Institution, meaning we could actually do this for our students without cost.  I'm looking into it more, but we might be able to move in this direction...

**Comment by cindyledev** [comment]
> Seneca is now an AWS Educate Institution, meaning we could actually do this for our students without cost. I'm looking into it more, but we might be able to move in this direction...

Wow! That's amazing. I just had to verify it and Seneca is indeed on the [list](https://s3.amazonaws.com/awseducate-list/AWS_Educate_Institutions.pdf)!

### According to https://aws.amazon.com/education/awseducate/aws-educate-faqs/
Students ages 18+ who enroll with an AWS Educate Starter account at a member institution receive $100 in AWS Promotional Credit and $30 in credit at a non-member institution.

Educators who enroll with an AWS account at a member institution receive $200 in AWS Promotional Credit and $75 in credit at a non-member institution.

Educators who enroll with an AWS Educate Starter account at a member institution receive $150 in AWS Promotional Credit and $50 in credit at a non-member institution.

These are the [supported services](https://awseducate-starter-account-services.s3.amazonaws.com/AWS_Educate_Starter_Account_Services_Supported.pdf) for AWS Educate Starter Accounts 

This is exciting, I'm gonna redo this doc for an ec2 instance that is AWS Educate Starter supported

**Comment by humphd** [comment]
I'm trying to understand the program better, and figure out how to use it broadly.  I'd love to be able to have remote VSCode sessions from containers running complex dev environments for all of my students.

I think solving this for Telescope dev in OSD700/DPS911 next semester might be a good starting point.

I'm very new to AWS, so this doc and your knowledge, @cindyledev, will be helpful.

I'll report back when I get more details from ITS at Seneca on what we can/can't do.

File: docs/aws-cloud9.md [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+- Minimum: `t2.medium (4 GiB RAM + 2 vCPU)`
+- Recommended: `t2.large (8 GiB RAM + 2 vCPU)`
+
+**Summary of Pricing**:
+
+- t2.medium costs \$0.0464 per hour
+- t2.large costs \$0.0928 per hour
+- 30GB Amazon Elastic Block Storage (EBS) costs \$3 per month
+
+**Example Usage**:
```
Comment (humphd) [inline docs/aws-cloud9.md:14]: So you're showing Cost Estimate Per Month?  Can we label it as such?

File: env.example [hunk 4]
```diff
@@ -4,6 +4,9 @@ NODE_ENV=development
 # PORT is the port used by the web server
 PORT=3000
 
+# AWS Public IPv4 address
+MY_PUBLIC_IP=(curl -s http://169.254.169.254/latest/meta-data/public-ipv4)
```
Comment (humphd) [inline env.example:8]: Will this break on machines that don't have curl?

File: env.example (hunk 1 above)
Comment (humphd) [reply env.example:22]: I'm not sure what the right approach to this is.

**Review by humphd — CHANGES_REQUESTED** [review]

File: docs/aws-cloud9.md (hunk 3 above)
Comment (cindyledev) [reply docs/aws-cloud9.md]: Good idea.

They're labeled now :)

File: env.example (hunk 4 above)
Comment (cindyledev) [reply env.example]: The AWS EC2 instances that are running Ubuntu and Amazon Linux have curl preinstalled. I figured out how to get the ip address so these lines are no longer needed

Comment (cindyledev) [reply env.example]: It won't let me comment on your third comment. But I made a script that will get the EC2 ip address and put it in the .env file so everything works now :)

**Comment by cindyledev** [comment]
I want to explore https://medium.com/@0xson/running-ubuntu-desktop-gui-aws-ec2-instance-on-windows-3d4d070da434 or something similar. This looks to be another cool way of using AWS EC2 instances

**Comment by birtony** [comment]
Hey @cindyledev, I can see you've worked on this pr for quite a while. Are you still planning on taking it further on?

**Comment by c3ho** [comment]
In terms of cost, it should be about ~100 for the month if you use a t2.large and leave it on for the month. The costs could be reduced a lot though if people remember to shut down running apps / shutdown the instance when they're not using it.

**Comment by cindyledev** [comment]
[This](https://github.com/Seneca-CDOT/telescope/pull/1298#discussion_r537700306) has been resolved but for some reason, Github will not let me flag the comment as resolved and will not let me comment on it.

**Review by humphd — APPROVED** [review]

File: docs/aws-cloud9.md [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+warning Error from chokidar (/home/ubuntu/environment/telescope/src/frontend/node_modules/@material-ui/icons): Error: ENOSPC: System limit for number of file watchers reached, watch '/home/ubuntu/environment/telescope/src/frontend/node_modules/@material-ui/icons/index.js'
+```
+
+It's okay, run
+
+```
+echo fs.inotify.max_user_watches=524288 | sudo tee -a /etc/sysctl.conf && sudo sysctl -p
+```
+
+### But I want to use Docker!!
```
Comment (chrispinkney) [inline docs/aws-cloud9.md:314]: 😄

**Review by chrispinkney — APPROVED** [review]

**Comment by chrispinkney** [comment]
Awesome. AWS is really slick, I'm really enjoying my time with it in my capstone project.

**Pull request merged**